src/vivarium/framework/engine.py

```diff
@@ -35,107 +33,136 @@
+        self.configuration.freeze()
 
-        self.lifecycle.set_state('post_setup')
-        self.builder.event.get_emitter('post_setup')(None)
+        self._lifecycle.set_state('post_setup')
+        self._builder.event.get_emitter('post_setup')(None)
 
     def initialize_simulants(self):
-        self.lifecycle.set_state('population_creation')
+        self._lifecycle.set_state('population_creation')
 
         pop_params = self.configuration.population
         # Fencepost the creation of the initial population.
-        self.clock.step_backward()
+        self._clock.step_backward()
         population_size = pop_params.population_size
         self.simulant_creator(population_size, population_configuration={'sim_state': 'setup'})
-        self.clock.step_forward()
+        self._clock.step_forward()
 
     def step(self):
-        _log.debug(self.clock.time)
+        _log.debug(self._clock.time)
         for event in self.time_step_events:
-            self.lifecycle.set_state(event)
-            self.time_step_emitters[event](self.population.get_population(True).index)
-        self.clock.step_forward()
+            self._lifecycle.set_state(event)
+            self.time_step_emitters[event](self._population.get_population(True).index)
+        self._clock.step_forward()
+
+    def run(self):
+        while self._clock.time < self._clock.stop_time:
+            self.step()
 
     def finalize(self):
-        self.lifecycle.set_state('simulation_end')
-        self.end_emitter(self.population.get_population(True).index)
+        self._lifecycle.set_state('simulation_end')
+        self.end_emitter(self._population.get_population(True).index)
+        unused_config_keys = self.configuration.unused_keys()
+        if unused_config_keys:
+            _log.debug("Some configuration keys not used during run: %s", unused_config_keys)
 
     def report(self):
-        return self.values.get_value('metrics')(self.population.get_population(True).index)
+        self._lifecycle.set_state('report')
+        metrics = self._values.get_value('metrics')(self._population.get_population(True).index)
+        _log.debug(pformat(metrics))
+        return metrics
 
     def add_components(self, component_list):
-        if self._setup:
-            raise VivariumError('Components cannot be added to a simulation once it is setup.')
-        self.component_manager.add_components(component_list)
+        """Adds new components to the simulation."""
+        # TODO: Set with life cycle restrictions.
+        if self._lifecycle.current_state.name != 'initialization':
+            raise LifeCycleError
+        self._component_manager.add_components(component_list)
+
+    def get_population(self, untracked: bool = True):
+        # TODO: Set with life cycle restrictions.
+        if self._lifecycle.current_state.name in ['bootstrap', 'initialization', 'setup', 'post_setup']:
+            raise LifeCycleError
```

Comment:
same comment re above. where are users supposed to go to know when they can call things since we're not telling them here